spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,181 @@
+InvokerInfo invoker = builder.build();
+----
+
+To call the `hello()` method through this invoker, assuming that `myService` is an instance of the bean, call:
+
+[source,java]
+----
+invoker.invoke(myService, new Object[] {"world"})
+----
+
+The return value is `"Hello world!"`.
+
+An implementation of the invoker above is equivalent to the following class:
+
+[source,java]
+----
+public class TheInvoker implements Invoker<MyService, String> {
+    public String invoke(MyService instance, Object[] arguments) {
+        return instance.hello((String) arguments[0]);
+    }
+}
+----
+
+[[invoker_builder]]
+=== Using `InvokerBuilder`
+
+`InvokerBuilder` can be obtained in build compatible extensions from `BeanInfo.createInvoker()`:
+
+[source,java]
+----
+public interface BeanInfo {
+    ...
+
+    InvokerBuilder<InvokerInfo> createInvoker(MethodInfo method);
+}
+----
+
+The target bean of the created invoker is the bean represented by the `BeanInfo` object.
+The target method of the created invoker is the method represented by the `MethodInfo` object passed to `createInvoker()`.
+
+Calling `InvokerBuilder.build()` produces an opaque token (`InvokerInfo`) that can be passed as a parameter to a `SyntheticBeanBuilder` or `SyntheticObserverBuilder` and materializes as an `Invoker` at application runtime.
+
+// TODO lookups, transformers, wrappers
+
+[[invoker_assignability]]
+=== Assignability for invokers
+
+For the purpose of `Invoker.invoke()`, assignability is defined as:
+
+* a non-`null` instance is assignable to the target bean when the class of the instance is a subclass of the bean class of the target bean;
+* a `null` argument is assignable to any parameter of the target method;
+* a non-`null` argument is assignable to a parameter of the target method when the class of the argument is assignable to the erasure of the declared type of the parameter, as defined below.
+
+If an argument class is denoted `A` and the erasure of the declared type of a parameter is denoted `P`, then:
+
+* `A` is assignable to a primitive type `P` when `A` is the primitive wrapper class of `P`;
+* `A` is assignable to a class or interface type `P` when:
+** `A` is a class or interface type and `A` is a subclass of `P`, or
+** `A` is an array type and `P` is `java.lang.Object`, `java.lang.Cloneable`, or `java.io.Serializable`;
+* `A` is assignable to an array type `P` when `A` is an array type and:
+** the component type of `P` is a primitive type and the component types of `A` and `P` are identical, or
+** the component type of `P` is not a primitive type and the component type of `A` is assignable to the component type of `P` according to these rules.
+
+NOTE: This is a reformulation of the JVM assignability rules as defined by the specification of the `checkcast` and `instanceof` instructions, with addition of primitive types.
+It is expected that CDI implementations do not implement these rules and rely on JVM runtime checks.
```

Comment:
The above rules are good but given this note and the fact that impls will rely on JVM checks anyway, do we need to explicitly state them? I mean, couldn't we just say they work the same JVM assignability rules with a link there?

Reply:
The thing is that they are not exactly the same, because we convert wrapper types to primitive types. It could probably be simplified, but I figured we should just specify this properly :shrug:

Reply:
I'm thinking we should allow widening conversion for primitive types, because that's what method handles (as well as reflection) do.

Reply:
> I'm thinking we should allow widening conversion for primitive types, because that's what method handles (as well as reflection) do.

That makes sense, +1

Reply:
I removed this section entirely, because we can just refer to method invocation conversions, as defined by JLS.